cloudify_cli/commands/upgrade.py

```diff
@@ -0,0 +1,128 @@
+########
+# Copyright (c) 2016 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+"""
+Handles 'cfy upgrade command'
+"""
+import yaml
+import time
+import json
+import tempfile
+from contextlib import contextmanager
+
+from cloudify_cli import utils
+from cloudify_cli import common
+from cloudify_cli import exceptions
+from cloudify_cli.logger import get_logger
+from cloudify_cli.commands import maintenance
+
+import fabric.api
+import fabric.context_managers
+
+MAINTENANCE_MODE_DEACTIVATED = 'deactivated'
+MAINTENANCE_MODE_ACTIVATING = 'activating'
+
+REMOTE_WORKFLOW_STATE_PATH = '/opt/manager/_workflow_state.json'
+
+logger = get_logger()
+
+
+def upgrade(validate_only,
+            skip_validations,
+            blueprint_path,
+            inputs,
+            install_plugins,
+            task_retries,
+            task_retry_interval,
+            task_thread_pool_size):
+
+    management_ip = utils.get_management_server_ip()
+    client = utils.get_rest_client(management_ip)
+
+    curr_status = client.maintenance_mode.status().status
+    if curr_status == MAINTENANCE_MODE_DEACTIVATED:
+        msg = "'Manager must be in maintenance mode for upgrade to run."
+        error = exceptions.CloudifyCliError(msg)
+        error.possible_solutions = [
+            "Activate maintenance mode using 'cfy maintenance-mode activate'"
+        ]
+        raise error
+    elif curr_status == MAINTENANCE_MODE_ACTIVATING:
+        _wait_for_maintenance(client)
+
+    env_name = 'manager-upgrade'
+    # init local workflow execution environment
+    env = common.initialize_blueprint(blueprint_path,
+                                      storage=None,
+                                      install_plugins=install_plugins,
+                                      name=env_name,
+                                      inputs=inputs)
+
+    if not skip_validations:
+        logger.info('Executing upgrade validation')
+        env.execute(workflow='execute_operation',
+                    parameters={'operation':
+                                'cloudify.interfaces.validation.creation'},
+                    task_retries=task_retries,
+                    task_retry_interval=task_retry_interval,
+                    task_thread_pool_size=task_thread_pool_size)
+
+    if not validate_only:
+        with open(inputs) as f:
```

Comment:
so, two things. one, inputs are not always a file. they can be multiple files, json, key value, etc. to handle these inputs, there is a util for that:

```python
inputs = utils.inputs_to_dict(inputs, 'inputs')
```

on top of that, I don't quite understand why you need inputs at all. While we might want to allow overriding all inputs, the `ssh_user` and `ssh_key_filename` are exposed using:

```python
ssh_key_path = utils.get_management_key()
username = utils.get_management_user()
```

This will save the user from having to provide those inputs when upgrading. Instead, they could be default just run `cfy upgrade -p ...`